### PR TITLE
feat(upgrade): upgrade database

### PR DESCRIPTION
**Pull request by vlastahajek**
This is 2nd PR of 4  for the upgrade feature.
It adds upgrading database logic.

- [X] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [X] Rebased/mergeable
- [X] Tests pass
- [ ] Documentation updated or issue created (provide link to issue/pr)

### CODE REVIEW

**Comment by vlastahajek** [comment]
It ready for comments, but I keep it in the draft,  cause I'm going to extend tests for db upgrade a bit.

File: cmd/influxd/upgrade/fs.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+		return
+	}
+
+	return
+}
+
+// CopyDir recursively copies a directory tree, attempting to preserve permissions.
+// Source directory must exist, destination directory must *not* exist.
+// Symlinks are ignored and skipped.
+// dirRenameFunc allows renaming  directories. It must return orig name, if renaming is not done..
```
Comment (stuartcarnie) [inline cmd/influxd/upgrade/fs.go:70]:
```suggestion
// dirRenameFunc is a mapping function that transforms path to a new name. Returning the path specifies the directory should not be renamed.
```

File: cmd/influxd/upgrade/fs.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+// dirFilterFunc allows filtering out directories. Returned true means skipping the dir.
```
Comment (stuartcarnie) [inline cmd/influxd/upgrade/fs.go:71]:
```suggestion
// dirFilterFunc ignores all directories where dirFilterFunc(path) is true. Passing nil for dirFilterFunc includes all directories.
```

File: cmd/influxd/upgrade/fs.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+// fileFilterFunc allows filtering out files. Returned true means skipping the file.
```
Comment (stuartcarnie) [inline cmd/influxd/upgrade/fs.go:72]:
```suggestion
// fileFilterFunc ignores all files where fileFilterFunc(path) is true. Passing nil for fileFilterFunc includes all files.
```

File: pkg/fs/fs.go [hunk 4]
```diff
@@ -15,3 +17,31 @@ func newFileExistsError(path string) FileExistsError {
...
+	case size > GB:
+		return fmt.Sprintf("%.1fGB", float64(size)/GB)
+	case size > MB:
+		return fmt.Sprintf("%.1fMB", float64(size)/MB)
+	case size > KB:
+		return fmt.Sprintf("%.1fKB", float64(size)/KB)
+	default:
+		return fmt.Sprintf("%dB", size)
+	}
+}
```
Comment (stuartcarnie) [inline pkg/fs/fs.go:47]: We already have a dependency on the https://github.com/dustin/go-humanize package, so you should be able to use `humanize.Bytes`:

https://pkg.go.dev/github.com/dustin/go-humanize#Bytes

File: pkg/fs/fs_unix.go [hunk 5]
```diff
@@ -5,6 +5,8 @@ package fs
 import (
 	"os"
 	"syscall"
+
+	syscall2 "golang.org/x/sys/unix"
```
Comment (stuartcarnie) [inline pkg/fs/fs_unix.go:9]: This should be imported as `unix`

**Review by stuartcarnie — CHANGES_REQUESTED** [review]
Looking good! 

Can you verify that `goimports` has been run across all the files, as I noticed some inconsistencies in formatting.

File: cmd/influxd/upgrade/database.go [hunk 6]
```diff
@@ -2,12 +2,183 @@ package upgrade
...
+	if err != nil {
+		return nil, fmt.Errorf("error getting size of %s: %w", v1opts.walDir, err)
+	}
+	size += size2
+	v2dir := filepath.Dir(v2opts.boltPath)
+	diskInfo, err := fs2.DiskUsage(v2dir)
+	if err != nil {
+		return nil, fmt.Errorf("error getting info of disk %s: %w", v2dir, err)
+	}
+	if options.verbose {
```
[the remark this answers is not in the document]
Comment (vlastahajek) [reply cmd/influxd/upgrade/database.go:49]: Ok

File: cmd/influxd/upgrade/database.go [hunk 7]
```diff
@@ -2,12 +2,183 @@ package upgrade
 
 import (
 	"context"
+	"fmt"
+	"path/filepath"
+
 	"github.com/influxdata/influxdb/v2"
-	"github.com/influxdata/influxdb/v2/kit/errors"
+	fs2 "github.com/influxdata/influxdb/v2/pkg/fs"
```
[the remark this answers is not in the document]
Comment (vlastahajek) [reply cmd/influxd/upgrade/database.go:9]: No. The local fs.go was previously in the `upgrade/fs` package and it was imported as `fs`. After moving to the same folder, `pkg/fs` stayed as `fs2`.

File: cmd/influxd/upgrade/database.go [hunk 8]
```diff
@@ -2,12 +2,183 @@ package upgrade
...
+	// read each database / retention policy from v1.meta and create bucket db-name/rp-name
+	// create database in v2.meta
+	// copy shard info from v1.meta
+	for _, db := range v1.meta.Databases() {
+		if db.Name[0] == '_' {
+			if options.verbose {
+				log.Info("Skipping internal ",
+					zap.String("database", db.Name))
+			}
+			continue
```
[the remark this answers is not in the document]
Comment (vlastahajek) [reply cmd/influxd/upgrade/database.go:67]: Good point! 👍 
I wasn't sure if there could be other internal databases, that would also start with '_'. But after testing with 1.8.1 I see user can create a database starting with '_'.

File: cmd/influxd/launcher/launcher_helpers.go [hunk 9]
```diff
@@ -48,6 +48,9 @@ type TestLauncher struct {
 	Stdin  bytes.Buffer
 	Stdout bytes.Buffer
 	Stderr bytes.Buffer
+
+	// Flag to act as standard server: disk store, no-e2e testing flag
+	RealServer bool
```
[the remark this answers is not in the document]
Comment (vlastahajek) [reply cmd/influxd/launcher/launcher_helpers.go:53]: Ok

File: cmd/influxd/upgrade/database.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (stuartcarnie) [reply cmd/influxd/upgrade/database.go:50]: Ok, that is great – my intent is that `verbose` is the default option if the user does not specify anything.

File: cmd/influxd/launcher/launcher_helpers.go [hunk 10]
```diff
@@ -70,6 +73,13 @@ func NewTestLauncher(flagger feature.Flagger) *TestLauncher {
 	return l
 }
 
+// NewTestLauncherServer returns a new instance of TestLauncher configured as real server (disk store, no e2e flag).
+func NewTestLauncherServer(flagger feature.Flagger) *TestLauncher {
+	l := NewTestLauncher(flagger)
+	l.realServer = true
+	return l
+}
+
```
Comment (stuartcarnie) [inline cmd/influxd/launcher/launcher_helpers.go:82]: 👏

**Review by stuartcarnie — APPROVED** [review]
Great! We're good to go with this one too 👏 👏

**Pull request merged**